corehq/form_processor/parsers/form.py

```diff
@@ -10,75 +12,94 @@
+        if raised:
+            # best effort: re-raise first error
+            six.reraise(*raised[0])
 
 
 class FormProcessingResult(object):
 
-    def __init__(self, submitted_form, existing_duplicate=None):
+    def __init__(self, submitted_form):
         self.submitted_form = submitted_form
-        self.existing_duplicate = existing_duplicate
         self.interface = FormProcessorInterface(self.submitted_form.domain)
 
-    def _get_form_lock(self, form_id):
-        return self.interface.acquire_lock_for_xform(form_id)
-
     def get_locked_forms(self):
-        if self.existing_duplicate:
-            # Lock docs with their original ID's (before they got switched during deprecation)
-            old_id = self.existing_duplicate.form_id
-            new_id = self.submitted_form.form_id
-            assert old_id != new_id, 'Expecting form IDs to be different'
-            return MultiLockManager([
-                LockManager(self.submitted_form, self._get_form_lock(new_id)),
-                LockManager(self.existing_duplicate, self._get_form_lock(old_id)),
-            ])
-        else:
-            return MultiLockManager([
-                LockManager(self.submitted_form, self._get_form_lock(self.submitted_form.form_id))
-            ])
-
-
-class LockedFormProcessingResult(FormProcessingResult):
-
-    def __init__(self, submitted_form):
-        super(LockedFormProcessingResult, self).__init__(submitted_form)
-        assert submitted_form.is_normal
-        self.lock = self._get_form_lock(submitted_form.form_id)
-
-    def get_locked_forms(self):
-        return MultiLockManager([LockManager(self.submitted_form, self.lock)])
+        """Get a context manager whose context is a list of locked forms
+
+        Form locks have been acquired by the time this method returns.
+        Call `__exit__` on the returned context manager to release the
+        locks.
+        """
+        xform = self.submitted_form
+        xform_id = xform.form_id
+        manager = FormLockManager([xform], self.interface)
+        manager.acquire_lock(xform_id)
+        try:
+            if self.interface.is_duplicate(xform_id):
+                new_form, dup_form = _handle_id_conflict(xform, xform.domain)
+                new_id = new_form.form_id
+                if dup_form:
+                    old_id = dup_form.form_id
+                    assert old_id != new_id, 'Expecting form IDs to be different'
+                    manager.context = [new_form, dup_form]
+                    if xform_id == new_id:
+                        manager.acquire_lock(old_id)
+                    elif xform_id == old_id:
+                        manager.acquire_lock(new_id)
+                    else:
+                        # unexpected, should never get here
+                        assert False, (xform_id, new_id, old_id)
+                elif new_id != xform_id:
+                    manager.context = [new_form]
+                    manager.acquire_lock(new_id)
```

Comment:
Do we even need a lock here? In this case the `new_id` has just been assigned to the form on the server so unless there's a UUID clash we shouldn't need to worry about another form being submitted with the same ID.

Reply:
The same applies to the code above.

Reply:
Good point. I just tried to stick to what the old code was doing. We may be able to remove these locks in the future, but I don't think they will hurt anything if they're released reliably.

Reply:
e8c86fc